shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -619,6 +630,21 @@ - (UITextRange*)rangeEnclosingPosition:(UITextPosition*)position
 }
 
 - (UITextRange*)lineEnclosingPosition:(UITextPosition*)position {
+  // TODO(hellohuanlin): Remove iOS 17 check. The same logic should apply to older versions too.
+  if (@available(iOS 17.0, *)) {
+    // The end of document (with backward affinity) should not be part of any line, since end is
+    // exclusive. This is to fix 2 bugs:
+    // 1. Keyboard freezes when switching languages:
+    // https://github.com/flutter/flutter/issues/134716
+    // 2. Auto correction candidate menu does not show up in iOS 17:
+    // https://github.com/flutter/flutter/issues/132594
+    // The end of document with forward affinity should still return the last line, which is used
+    // by voice control's delete line command.
+    if ([position isEqual:[_textInputView endOfDocument]]) {
+      return nil;
```

Comment:
What would happen if this returns the last line?

Reply:
If we return the last line, we will see the 2 bugs linked in my description.

Reply:
I see, if `endOfDocument` is exclusive and does return `(text.length, upstream)`, then it is contradicting with how flutter handle text position.

Reply:
`(text.length, downstream)` is a valid cursor position as well. In most cases `(text.length, upstream)` and `(text.length, downstream)` refer to the same cursor positoin, but when there's bidi text at the end of the text they refer to different locations.

Reply:
I tried in `UITextField`. It's returning `7F` for a string that's 7 code units long. So it's inclusive I think.

Reply:
I can play around with it more tomorrow, but afaik apple's [sample project](https://developer.apple.com/library/archive/samplecode/SimpleTextInput/Listings/SimpleTextInput_APLEditableCoreTextView_m.html#//apple_ref/doc/uid/DTS40010633-SimpleTextInput_APLEditableCoreTextView_m-DontLinkElementID_8) uses exclusive endOfDocument. (It does not use affinity tho)

Reply:
Yeah cursor position and text position have different end positions. I'm not sure which one `UITextField` is using.